rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,311 @@
+use std::collections::HashMap;
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+
+use rusoto_core::{Region, RusotoError};
+use rusoto_dynamodb::{
+    AttributeValue, DynamoDb, DynamoDbClient, GetItemError, GetItemInput, PutItemError,
+    PutItemInput,
+};
+use uuid::Uuid;
+
+const PARTITION_KEY_NAME: &str = "key";
+const PARTITION_KEY_VALUE: &str = "delta-rs";
+const OWNER_NAME: &str = "ownerName";
+const RECORD_VERSION_NUMBER: &str = "recordVersionNumber";
+const IS_RELEASED: &str = "isReleased";
+const LEASE_DURATION: &str = "leaseDuration";
+
+mod expressions {
+    pub const ACQUIRE_LOCK_THAT_DOESNT_EXIST: &str = "attribute_not_exists(#pk)";
+
+    pub const PK_EXISTS_AND_IS_RELEASED: &str = "attribute_exists(#pk) AND #ir = :ir";
+
+    pub const PK_EXISTS_AND_RVN_IS_THE_SAME: &str = "attribute_exists(#pk) AND #rvn = :rvn";
+}
+
+mod vars {
+    pub const PK_PATH: &str = "#pk";
+    pub const RVN_PATH: &str = "#rvn";
+    pub const RVN_VALUE: &str = ":rvn";
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+}
+
+pub struct DynamoLock {
+    client: DynamoDbClient,
+    table_name: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    owner_name: String,
+    record_version_number: String,
+    lease_duration: u64,
+    is_released: bool,
+
+    lookup_time: u128,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    #[error("Condition failed: {0}")]
+    ConditionalCheckFailed(String),
+
+    #[error("DynamoDB item has invalid schema")]
+    InvalidItemSchema,
+
+    #[error("Could not acquire lock for {0} sec")]
+    TimedOut(u64),
+
+    #[error("Get item error: {0}")]
+    GetItemError(RusotoError<GetItemError>),
+
+    #[error("Put item error: {0}")]
+    PutItemError(RusotoError<PutItemError>),
+}
+
+impl DynamoLock {
+    pub fn new() -> Self {
+        let client = DynamoDbClient::new(Region::default());
```

Comment:
would be better to refactor and reuse the same provider construction logic used in S3 client: https://github.com/delta-io/delta-rs/blob/main/rust/src/storage/s3.rs#L59-L74.

Reply:
@houqp Btw, does it make sense to share the same client as well? (`rusoto_core::Client`)

Reply:
probably better to keep them separate unless we get performance benefit out of that.